lightning: add docs for local backend (#3660)
pingcap/docs: tidb-lightning/tidb-lightning-backends.md

---
title: TiDB Lightning Import Modes
summary: Learn the different import modes of TiDB Lightning.
aliases: ['/docs/dev/tidb-lightning/tidb-lightning-tidb-backend/','/docs/dev/reference/tools/tidb-lightning/tidb-backend/','/tidb/dev/tidb-lightning-tidb-backend']
---

# TiDB Lightning Import Modes

> TiDB Lightning supports three backends: Importer-backend (default), Local-backend, and TiDB-backend. It determines how `tidb-lightning` delivers data into the target cluster.

Reviewer: ```suggestion
TiDB Lightning supports three backends: Importer-backend (default), Local-backend, and TiDB-backend, which determine how `tidb-lightning` imports data into the target cluster.
```
I think "import" is a more accurate verb here.

> TiDB Lightning supports three backends: Importer-backend (default), Local-backend, and TiDB-backend. It determines how `tidb-lightning` delivers data into the target cluster.

Reviewer: ```suggestion
TiDB Lightning supports three backends (import modes): Importer-backend (default), Local-backend, and TiDB-backend. The backend determines how `tidb-lightning` delivers data into the target cluster.
```
Please help confirm in this context: backends = import modes? @TomShawn

Author: @glorv PTAL

Reviewer 2: Yes, I think `import mode` or something similar is a better way to express these three different kind of running type. `backend` is not easy for user to understand.

BTW, because the former version of this doc use `backend` to describe the difference, so when add `local` mode, I just followed it.

@kennytm  PTAL

Reviewer 3: @glorv @TomShawn @yikeke  "import mode" is already used together with "normal mode". so it is bad to reuse the same terminology for something entirely different (backend).

Reviewer: Makes sense. @kennytm I found the two terms in our glossary: [back end](https://docs.pingcap.com/tidb/dev/tidb-lightning-glossary#back-end) and [import mode](https://docs.pingcap.com/tidb/dev/tidb-lightning-glossary#import-mode). I think it'll be clearer if we give a reference link to this terminology at the beginning. @TomShawn 

Also, I think we should change "导入模式" to "后端" as it was introduced in [this PR](https://github.com/pingcap/docs-cn/pull/3979/files#diff-321248b8515b109cff5da454ea863ee7R2). @TomShawn

Author: Done in https://github.com/pingcap/docs/pull/3660/commits/16cb7f55cb2cc43dff87b82970498f4d49c02685
@yikeke @kennytm PTAL

> 

Reviewer: How about putting the following three items to a list?

Author: Done

The **Importer-backend** (default) requires `tidb-lightning` to first encode the SQL or CSV data into KV pairs, and relies on the external `tikv-importer` program to sort these KV pairs and ingest directly into the TiKV nodes.

The **Local-backend**: `tidb-lightning` first encodes data into key-value pairs, sorts and stores them in a local temporary directory, and writes these key-value pairs to each TiKV node in batches. Then, TiKV ingests these key-value pairs into the cluster. The implementation of Local-backend is the same with that of Importer-backend but does not rely on the external `tikv-importer` component.

> The **TiDB-backend** requires `tidb-lightning` to encode these data into SQL `INSERT` statements, and has these statements executed directly on the TiDB node.

Reviewer: Seems not consistent with the original latest version: https://docs.pingcap.com/zh/tidb/dev/tidb-lightning-backends?

Author: I put "Importer-backend" before "Local-backend" because it is still the default mode. @glorv What do you think?

> | Back end | Local-backend | Importer-backend | TiDB-backend |

Reviewer: ```suggestion
| Backend | Local-backend | Importer-backend | TiDB-backend |
```

|:---|:---|:---|:---|
| Speed | Fast (~500 GB/hr) | Fast (~300 GB/hr) | Slow (~50 GB/hr) |
| Resource usage | High | High | Low |
| Network bandwidth usage| High | Medium | Low |
| ACID respected while importing | No | No | Yes |
| Target tables | Must be empty | Must be empty | Can be populated |

> | External component needed | No | `tikv-importer` | No |

Reviewer: ```suggestion
| Additional component required | No | `tikv-importer` | No |
```

| TiDB versions supported | >= v4.0.0 | All | All |

## How to choose the backend modes

- If the target cluster of data import is v4.0 or later versions, consider using the Local-backend mode first, which is easier to use and has higher performance than that of the other two modes.
- If the target cluster of data import is v3.x or earlier versions, it is recommended to use the Importer-backend mode.
- If the target cluster of data import is in the online production environment, or if the target table of data import already has data on it, it is recommended to use the TiDB-backend mode.

## TiDB Lightning Local-backend

> The Local-backend feature was introduced in TiDB v4.0.3. TiDB Lightning only of v4.0.3 or later versions has this feature. In addition, Local-backend only supports TiDB cluster of v4.0.0.

Reviewer: ```suggestion
The Local-backend feature was introduced in TiDB v4.0.3. TiDB Lightning only of v4.0.3 or later versions has this feature. In addition, Local-backend only supports TiDB clusters of v4.0.0 or above.
```
I'm confused because the first two sentences seem to conflict with the last sentence. So can a v4.0.1 TiDB cluster use the Local-backend feature?

Author: The original Chinese is `Local-backend 特性在 TiDB v4.0.3 发布，v4.0.3 及以上的 TiDB Lightning 才包含此特性。另外 Local-backend 只支持 v4.0.0 以上的集群。` @glorv Could you answer the question above?

Reviewer 2: We release `local backend` in reversion v4.0.3. But it can support restore data to tidb clusters of v4.0.0 or above(include v4.0.0~4.0.2).
I want to express only with tidb-lightning >= v4.0.3  can we use this feature. But It's ok for all tidb clusters >= v4.0.0

Reviewer 2: This conflict is caused by the fact that we always release tidb(tidb/tikv/pd ...) and its toolset (lightning/br...) together, but the toolset is often compatible with multi versions of tidb

Reviewer: ```suggestion
The Local-backend feature is introduced to TiDB Lightning since TiDB v4.0.3. But you can use this feature to import data to TiDB clusters of v4.0.0 or above.
```
How about saying in this way?

Reviewer: And for Chinese: `自 TiDB 4.0.3 版本起，TiDB Lightning 引入了 Local-backend 特性。该特性支持导入数据到 v4.0.0 以上的 TiDB 集群。` @TomShawn @glorv

### Deployment for Local-backend

To deploy TiDB Lightning in the Local-backend mode, see [TiDB Lightning Deployment](/tidb-lightning/deploy-tidb-lightning.md).

## TiDB Lightning TiDB-backend

### Deployment for TiDB-backend

> When using the TiDB-backend, you no longer need `tikv-importer`. Compared with the [standard deployment procedure](/tidb-lightning/deploy-tidb-lightning.md), the TiDB-backend deployment has the following two differences:

Reviewer: ```suggestion
When using the TiDB-backend, deploying `tikv-importer` is not necessary. Compared with the [standard deployment procedure](/tidb-lightning/deploy-tidb-lightning.md), the TiDB-backend deployment has the following two differences:
```

> * Steps involving `tikv-importer` can all be skipped.

Reviewer: ```suggestion
* All steps involving `tikv-importer` can be skipped.
```

> * The configuration must be changed to indicate the TiDB-backend is used.

Reviewer: ```suggestion
* The configuration must be changed to declare that the TiDB-backend is used.
```

---
title: TiDB Lightning Backends
summary: Learn the backends of TiDB Lightning.
aliases: ['/docs/dev/tidb-lightning/tidb-lightning-tidb-backend/','/docs/dev/reference/tools/tidb-lightning/tidb-backend/','/tidb/dev/tidb-lightning-tidb-backend']
---

# TiDB Lightning Backends

> The backend determines how `tidb-lightning` imports data into the target cluster.

Reviewer: ```suggestion
The backend determines how TiDB Lightning imports data into the target cluster.
```

+ [Importer-backend](#tidb-lightning-importer-backend) (default)
+ [Local-backend](#tidb-lightning-local-backend)
+ [TiDB-backend](#tidb-lightning-tidb-backend)

> The backend determines how `tidb-lightning` imports data into the target cluster.

Reviewer: To be user-friendly: State what "backend" can do first, and say what backends Lightning supports.

Author: Done

> The **Importer-backend** (default) requires `tidb-lightning` to first encode the SQL or CSV data into KV pairs, and relies on the external `tikv-importer` program to sort these KV pairs and ingest directly into the TiKV nodes.

Reviewer: ```suggestion
The **Importer-backend** (default): `tidb-lightning` first encodes the SQL or CSV data into KV pairs, and relies on the external `tikv-importer` program to sort these KV pairs and ingest directly into the TiKV nodes.
```

The **Local-backend**: `tidb-lightning` first encodes data into key-value pairs, sorts and stores them in a local temporary directory, and writes these key-value pairs to each TiKV node in batches. Then, TiKV ingests these key-value pairs into the cluster. The implementation of Local-backend is the same with that of Importer-backend but does not rely on the external `tikv-importer` component.

> The **TiDB-backend** requires `tidb-lightning` to encode these data into SQL `INSERT` statements, and has these statements executed directly on the TiDB node.

Reviewer: ```suggestion
The **TiDB-backend**: `tidb-lightning` first encodes these data into SQL `INSERT` statements, and has these statements executed directly on the TiDB node.
```
